content/docs/futures/getting_asynchronous.md

```diff
@@ -0,0 +1,174 @@
+TCP socket has been established.
+
+```rust
+# #![deny(deprecated)]
+extern crate tokio;
+extern crate bytes;
+#[macro_use]
+extern crate futures;
+
+use tokio::io::AsyncWrite;
+use tokio::net::{TcpStream, tcp::ConnectFuture};
+use bytes::{Bytes, Buf};
+use futures::{Future, Async, Poll};
+use std::io::{self, Cursor};
+
+enum HelloWorld {
+    Connecting(ConnectFuture),
+    Connected(TcpStream, Cursor<Bytes>),
+}
+
+impl Future for HelloWorld {
+    type Item = ();
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<(), io::Error> {
+        use self::HelloWorld::*;
+
+        loop {
+            let socket = match *self {
+                Connecting(ref mut f) => {
+                    try_ready!(f.poll())
+                }
+                Connected(ref mut socket, ref mut data) => {
+                    while data.has_remaining() {
+                        try_ready!(socket.write_buf(data));
+                    }
+
+                    return Ok(Async::Ready(()));
+                }
+            };
+
+            let data = Cursor::new(Bytes::from_static(b"hello world"));
+            *self = Connected(socket, data);
+        }
+    }
+}
+
+fn main() {
+    let addr = "127.0.0.1:1234".parse().unwrap();
+    let connect_future = TcpStream::connect(&addr);
+    let hello_world = HelloWorld::Connecting(connect_future);
+# let hello_world = futures::future::ok::<(), ()>(());
+
+    // Run it
+    tokio::run(hello_world)
+}
+```
+
+This future is represented as an enumeration of states:
+
+1. Connecting
+2. Writing "hello world" to the socket.
+
+The future starts in the connecting state with an inner future of type
+[`ConnectFuture`]. It repeatedly polls this future until the socket is returned.
+The state is then transitioned to `Connected`.
+
+From the `Connected` state, the future writes data to the socket. This is done
+with the [`write_buf`] function. I/O functions are covered in more detail in the
+[next section][io_section]. Briefly, [`write_buf`] is a non-blocking function to
+write data to the socket. If the socket is not ready to accept the write,
+`NotReady` is returned. If some data (but not necessarily all) was written,
+`Ready(n)` is returned, where `n` is the number of written bytes. The cursor is
+also advanced.
+
+Once in the `Connected` state, the future must loops as long as there is data
+left to write. Because [`write_buf`] is wrapped with `try_ready!()`, when
+[`write_buf`] returns `NotReady`, our `poll` function returns with `NotReady`.
+
+At some point in the future, our `poll` function is called again. Because it is
+in the `Connected` state, it jumps directly to writing data.
+
+**Note** the loops, they are important. Many future implementations contain
+loops. These loops are necessary because `poll` cannot return until either all
+the data is written to the socket, or an inner future ([`ConnectFuture`] or
+[`write_buf`]) returns `NotReady`.
```

Comment:
Again, may be worth linking to a section with a longer explanation of what happens if you mess up the contract